server/resources/resource_management.go

```diff
@@ -834,45 +835,41 @@ func DeleteResource(context middleware.Context,
 		return fmt.Errorf("No environment for schema")
 	}
 	auth := context["auth"].(schema.Authorization)
-	policy, err := loadPolicy(context, "delete", strings.Replace(resourceSchema.GetSingleURL(), ":id", resourceID, 1), auth)
-	if err != nil {
-		return err
-	}
-	context["policy"] = policy
 
 	var resource *schema.Resource
 	var fetchErr error
 
+	if err := extension.HandleEvent(context, environment, "pre_delete", resourceSchema.ID); err != nil {
+		return err
+	}
+
 	if errPreTx := db.Within(dataStore, func(preTransaction transaction.Transaction) error {
-		tenantIDs := policy.GetTenantIDFilter(schema.ActionDelete, auth.TenantID())
-		filter := transaction.IDFilter(resourceID)
-		if tenantIDs != nil {
-			filter["tenant_id"] = tenantIDs
+		resource, fetchErr = fetchResourceForAction(schema.ActionDelete, auth, resourceID, resourceSchema, preTransaction, context)
+		if fetchErr != nil {
+			switch fetchErr {
+			case transaction.ErrResourceNotFound:
+				_, err := fetchResourceForAction(schema.ActionRead, auth, resourceID, resourceSchema, preTransaction, context)
```

Comment:
Just an idea. Can we verify "resource exists but no privilege for the action" in one fetch operation? It's just a wasteful cost to access DB just for error code tweak.

Reply:
in theory, we can - we have to fetch resource without any filter and if it exists we can manually check (using reflection) that it meets the policy. in practice, it involves much work to do such check (consider new policy allowing complex conditions with 'or'/'and') and this is not the common case when user tries to delete non-existing resource or resource that he's not allowed to delete.

Reply:
got it